File: dodam-core/src/main/java/b1nd/dodamcore/outsleeping/repository/OutSleepingRepository.java
package b1nd.dodamcore.outsleeping.repository;

import b1nd.dodamcore.member.domain.entity.Student;
import b1nd.dodamcore.outsleeping.domain.entity.OutSleeping;
import b1nd.dodamcore.outsleeping.domain.exception.OutSleepingNotFoundException;
import org.springframework.data.jpa.repository.EntityGraph;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.time.LocalDate;
import java.util.List;

@Repository
public interface OutSleepingRepository extends JpaRepository<OutSleeping, Long> {

    default OutSleeping getById(Long id) {
        return findById(id)
                .orElseThrow(OutSleepingNotFoundException::new);
    }

    @EntityGraph(attributePaths = {"student.member"})
    @Query("select o from OutSleeping o where :date between o.startAt and o.endAt")
    List<OutSleeping> findByDate(@Param("date") LocalDate date);

    @EntityGraph(attributePaths = {"student.member"})
    List<OutSleeping> findByEndAtGreaterThanEqual(@Param("endAt") LocalDate endAt);

    @EntityGraph(attributePaths = {"student.member"})
    List<OutSleeping> findByStudentAndEndAtGreaterThanEqual(Student student, LocalDate now);

}

File: services/service-out-sleeping/src/main/kotlin/com/b1nd/dodamdodam/outsleeping/domain/outsleeping/repository/OutSleepingRepository.kt
package com.b1nd.dodamdodam.outsleeping.domain.outsleeping.repository

import com.b1nd.dodamdodam.outsleeping.domain.outsleeping.entity.OutSleepingEntity
import com.b1nd.dodamdodam.outsleeping.domain.outsleeping.enumeration.OutSleepingStatus
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import java.time.LocalDate
import java.util.UUID

interface OutSleepingRepository : JpaRepository<OutSleepingEntity, Long> {

    fun findAllByUserId(userId: UUID): List<OutSleepingEntity>

    fun findAllByStartAtLessThanEqualAndEndAtGreaterThanEqual(
        endDate: LocalDate,
        startDate: LocalDate
    ): List<OutSleepingEntity>

    fun findAllByStartAtLessThanEqualAndEndAtGreaterThanEqual(
        endDate: LocalDate,
        startDate: LocalDate,
        pageable: Pageable
    ): Page<OutSleepingEntity>

    fun findAllByStatusAndStartAtLessThanEqualAndEndAtGreaterThanEqual(
        status: OutSleepingStatus,
        endDate: LocalDate,
        startDate: LocalDate
    ): List<OutSleepingEntity>
}
